### PR TITLE
Complex scalar, vector, and matrix types

### PR DESCRIPTION
**Pull request by bob-carpenter**
This is a proposal to add complex scalar, vector, and matrix types to the Stan language.

### CODE REVIEW

**Comment by SteveBronder** [comment]
Linking to the rendered version below

https://github.com/stan-dev/design-docs/blob/complex-design/designs/0030-complex-numbers.md

File: designs/0030-complex-numbers.md [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+
+Lower and upper bounds could be construed as bounds on absolute value (or modulus) of a complex number, with
+```
+complex<lower = 1> z;
+```
+requiring `modulus(z) > 1`, where $\mod x + yi = \sqrt{x^2 + y^2}.$  Upper bounds and mixed upper and lower bounds work the same way.  The bounds are real valued.
+
+Lower and upper bound constraints can be applied to any of the complex types.
+
+This feature is included for completeness, but marked "optional" because it is not clear it will be useful as realistic complex constraints are typically multivariate.
```
Comment (WardBrian) [inline designs/0030-complex-numbers.md:71]: This was not implemented

File: designs/0030-complex-numbers.md [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+
+A complex number is constructed from a pair of real numbers, one representing the real component and one the imaginary component.  The proposal is to have a function do that work.
+
+```
+real r;
+real i;
+complex z = complex(r, i);
+```
+
+This overloads the type name in the same way as a C++ constructor.  `to_complex` seems too wordy and `cmplx` too cryptic.
```
Comment (WardBrian) [inline designs/0030-complex-numbers.md:84]: In the end, `to_complex` was used

File: designs/0030-complex-numbers.md [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+
+### Accessor lvalues
+
+The real and imaginary components of a complex number may be accessed as follows.
```
Comment (WardBrian) [inline designs/0030-complex-numbers.md:88]: While the following section would still be cool, the current implementation differs, using

`get_real(z)` and `get_imag(z)`. I don't believe these can be used as lvalues

File: designs/0030-complex-numbers.md [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+
+Stan programs are converted to C++ class definitions by the stanc3 transpiler.   There are two components in this compiled class that.  This section covers data input and parameter output.  User-facing I/O is handled by the interfaces.
+
+### Data input format
+
+The C++ model class uses an object of class `var_context` to input variables declared in the data block.  This object needs to be extended to read complex number objects.  To read a complex number, we must read a sequence of two real numbers to pass to the C++ `complex<double>` constructor.
+
+### Parameter output format
+
+Parameters are output as an array.  Complex numbers will be sequenced with their real components before their imaginary components.  The output names for a complex variable `z` will be `z.r` and `z.i` for the real and imaginary components.
```
Comment (WardBrian) [inline designs/0030-complex-numbers.md:265]: These ended up as `z.real` and `z.imag`, not `r` and `i`

File: designs/0030-complex-numbers.md [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+but that's going to be far too verbose.
+
+### R dump format input
+
+We can adopt the standard R dump format for complex number output, which separates the real and imaginary component with the 
+
+```
+z1 <- 3-1.5i
+z2 <- 2.3+4.7i
+```
```
Comment (WardBrian) [inline designs/0030-complex-numbers.md:298]: This has not been implemented, but complex numbers can be read in with something like 

`z1 <- c(3, 1.5)`

similar to JSON

File: designs/0030-complex-numbers.md [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+
+## Rationale and alternatives
+
+Most of these proposals are just applying the simplest possible thing that is mathematically and computationally coherent.  For example, it's natural to promote `real` to `complex` in mathematics, so we allow it in the language.  Similarly, it's natural to
```
Comment (WardBrian) [inline designs/0030-complex-numbers.md:302]: The final sentence here is incomplete

File: designs/0030-complex-numbers.md [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+
+### Container type alternative
+
+It might be tempting to try to decompose types as C++ does, for example using `matrix<real>` and `matrix<complex>` or even `complex matrix`, etc.  One obstacle to this is that we use `<...>` for constraints, as in `matrix<lower = 0>`.  And we want to start stacking constraints, as in `real<multiplier = 7, offset = 3.9><lower = 0>`.  But the real reason not to do this is that there's not much shared between the matrices other than constructors, getters, and setters.
+
+For example, complex numbers do *not* work with constraints.  It does not make sense to write `complex<lower = 0> z` or `complex<offset = 7>`.  
+
+### Full covariance
+
+The biggest issue is how much covariance do we want to support given that it's not yet fully supported for `int` and `real`.  For example, it'd be nice to allow assignment of `int[]` to `real[]` and of `real[]` to `complex[]`.
```
Comment (WardBrian) [inline designs/0030-complex-numbers.md:312]: Since the writing of this design this has become possible

File: designs/0030-complex-numbers.md [hunk 8]
```diff
@@ -0,0 +1,355 @@
...
+
+### Constructor alternatives
+
+We could use a simple pair structure instead of `complex(r, i)`, such as `{ r, i }`, `<r, i>`, or `[r, i]`, e.g.,
+```
+complex z = { r, i };
+```
+but these notations are already used for arrays, constraints, and row vectors respectively.
```
Comment (WardBrian) [inline designs/0030-complex-numbers.md:320]: This should discuss the literal syntax added later, `3+4i`

File: designs/0030-complex-numbers.md [hunk 9]
```diff
@@ -0,0 +1,355 @@
...
+
+For accessors, we could use method-like notation, such as `z.real()`, but Stan does not use this kind of object-oriented notation elsewhere and it is less clear that the result can be used an lvalue.
+
+### Offset and multiplier addition
+
+In addition to lower and upper bound constraints, it'd be possible to use complex offsets and multipliers, 
+```
+complex<offset = o, multiplier = m> z;
+```
+Here, the offset and multiplier could be complex scalars or real scalars or integers.
```
Comment (WardBrian) [inline designs/0030-complex-numbers.md:333]: Also not implemented, and in an above paragraph it is stated this would not make sense

**Review by WardBrian — COMMENTED** [review]
These comments are mainly to bring this document in line with what was implemented in https://github.com/stan-dev/stanc3/pull/940 before proceeding with the addition of complex matrices and vectors

File: designs/0030-complex-numbers.md (hunk 1 above)
Comment (bob-carpenter) [reply designs/0030-complex-numbers.md:71]: I think it's OK to leave it as an optional feature.  I'm not sure it's useful, so also feel free to cut.

File: designs/0030-complex-numbers.md (hunk 2 above)
Comment (bob-carpenter) [reply designs/0030-complex-numbers.md:84]: We should change this, then.

File: designs/0030-complex-numbers.md (hunk 3 above)
Comment (bob-carpenter) [reply designs/0030-complex-numbers.md:88]: Maybe when we get tuples?  So this should be updated with the . notation either removed or moved to a section marked "optional".

File: designs/0030-complex-numbers.md (hunk 4 above)
Comment (bob-carpenter) [reply designs/0030-complex-numbers.md:265]: At least it's only one line :-).  The doc should change.

File: designs/0030-complex-numbers.md (hunk 5 above)
Comment (bob-carpenter) [reply designs/0030-complex-numbers.md:298]: OK, then this also needs to be updated.

File: designs/0030-complex-numbers.md (hunk 6 above)
Comment (bob-carpenter) [reply designs/0030-complex-numbers.md:302]: delete.

File: designs/0030-complex-numbers.md (hunk 7 above)
Comment (bob-carpenter) [reply designs/0030-complex-numbers.md:312]: nice, so we can roll that into the proposal.

File: designs/0030-complex-numbers.md (hunk 8 above)
Comment (bob-carpenter) [reply designs/0030-complex-numbers.md:320]: Yes, please.

File: designs/0030-complex-numbers.md (hunk 9 above)
Comment (bob-carpenter) [reply designs/0030-complex-numbers.md:333]: Right.  We should just remove it.

**Review by bob-carpenter — COMMENTED** [review]
@WardBrian :  this was just comments at this point.  Do you want to make the changes and then I can approve?  Or would you rather I make the changes?

File: designs/0030-complex-numbers.md (hunk 1 above)
Comment (bob-carpenter) [reply designs/0030-complex-numbers.md:71]: Removed.

File: designs/0030-complex-numbers.md (hunk 2 above)
Comment (bob-carpenter) [reply designs/0030-complex-numbers.md:84]: fixed.

File: designs/0030-complex-numbers.md [hunk 10]
```diff
@@ -0,0 +1,328 @@
...
+real r = get_real(z);
+real i = get_imag(z);
+```
+
+These accessors do not work as lvalues. 
+
+
+### Equality
+
+Equality for scalars is defined component-wise, so that if `z1` and `z2` are are complex typed variables, then `z1 == z2` is true if `to_rea(z1) == to_real(z2)` and `to_real(z1) == to_real(z2)`.
```
Comment (WardBrian) [inline designs/0030-complex-numbers.md:98]:
```suggestion
Equality for scalars is defined component-wise, so that if `z1` and `z2` are are complex typed variables, then `z1 == z2` is true if `get_real(z1) == get_real(z2)` and `get_imag(z1) == get_imag(z2)`.
```

**Review by WardBrian — APPROVED** [review]
One typo/inconsistency but otherwise this looks good to me. Thanks

**Comment by bob-carpenter** [comment]
Thanks for catching that.  I made the update and pushed.

**Review by WardBrian — APPROVED** [review]
